ruamel-examples/parse_it.py

```diff
@@ -0,0 +1,237 @@
+        type_tags = {
+            "int": f"tag:{YAML_SPEC_TAG}:int",
+            "float": f"tag:{YAML_SPEC_TAG}:float",
+            "bool": f"tag:{YAML_SPEC_TAG}:bool",
+            "str": f"tag:{YAML_SPEC_TAG}:str",
+        }
+        if tag == "!ENV":
+            type_tag = type(value).__name__
+            actual_type = type_tags[type_tag]
+            return super().represent_scalar(actual_type, value, style, anchor=anchor)
+        return super().represent_scalar(tag, value, style, anchor=anchor)
+
+
+class EsmToolsLoader(ruamel.yaml.YAML):
+    def __init__(self, *args, **kwargs):
+        super().__init__(*args, **kwargs)
+        self.filename = None
+        self.add_comments = True
+        self.Constructor = ProvenanceConstructor
+        self.Constructor.add_constructor(None, ProvenanceConstructor.construct_scalar)
+        self.Constructor.add_constructor(
+            "tag:yaml.org,2002:bool", ProvenanceConstructor.construct_yaml_bool
+        )
+        self.Representer = EnvironmentRepresenter
+
+    def get_filename(self):
+        return self.filename
+
+    def set_filename(self, filename):
+        self.filename = filename
+
+    def load(self, stream):
+        self.set_filename(stream.name)
+        mapping_with_tuple_prov = super().load(stream)
+
+
+        def _extract_dict(tup, config=None, prov=False):
+            """
+            Iterative function to extract dictionary with or without provenance
+
+            Parameters
+            ----------
+            tup : tuple
+            config: dictionary holding either main config of provenance
+            prov: If main or provenance dict should be returned (True or False, default False)
+
+            Returns
+            -------
+            canfig: dictionary with or wihtout provenance
+
+            """
+            config = config if config else {}
+# unpack tuple
+            diction, (line,col) =  tup
+# check if diction is ruamel.yaml.comments.CommentedMap (of class dict)
+            if isinstance(diction, ruamel.yaml.comments.CommentedMap):
+                for (key,prov1),(value,prov2) in diction.items():
+                    if isinstance(value, ruamel.yaml.comments.CommentedMap):
+                        config[key] = _extract_dict((value,prov2), config, prov)
+                    elif isinstance(value, list):
+                        templist = []
+                        for (lkey,prov3) in value:
+                            if prov:
+                                templist.append(prov3)
+                            else:
+                                templist.append(lkey)
+                        config[key] = templist
+                    else:
+                        if prov:
+                            config[key] = prov1
+                        else:
+                            config[key] = value
+            else:
+                print("------------------------------------------------")
+            return (config)
```

Comment:
This function should return 2 dictionaries, one with the `data` and another one with the `provenance` information.